File: go/extensions/eip2612gassponsor/facilitator.go
package eip2612gassponsor

import (
	"encoding/json"
	"fmt"
	"regexp"
)

// ExtractEip2612GasSponsoringInfo extracts the EIP-2612 gas sponsoring info
// from a payment payload's extensions (raw JSON bytes).
//
// Returns the info if the extension is present and contains the required
// client-populated fields, or nil if not present.
func ExtractEip2612GasSponsoringInfo(extensions map[string]interface{}) (*Info, error) {
	if extensions == nil {
		return nil, nil
	}

	extRaw, ok := extensions[EIP2612GasSponsoring.Key()]
	if !ok {
		return nil, nil
	}

	// Marshal and unmarshal to get the extension structure
	extJSON, err := json.Marshal(extRaw)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal eip2612GasSponsoring extension: %w", err)
	}

	var ext Extension
	if err := json.Unmarshal(extJSON, &ext); err != nil {
		return nil, fmt.Errorf("failed to unmarshal eip2612GasSponsoring extension: %w", err)
	}

	// Marshal and unmarshal info to get the typed struct
	infoJSON, err := json.Marshal(ext.Info)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal eip2612GasSponsoring info: %w", err)
	}

	var info Info
	if err := json.Unmarshal(infoJSON, &info); err != nil {
		return nil, fmt.Errorf("failed to unmarshal eip2612GasSponsoring info: %w", err)
	}

	// Check that the client has populated the required fields
	if info.From == "" || info.Asset == "" || info.Spender == "" ||
		info.Amount == "" || info.Nonce == "" || info.Deadline == "" ||
		info.Signature == "" || info.Version == "" {
		return nil, nil
	}

	return &info, nil
}

// ExtractEip2612GasSponsoringInfoFromPayloadBytes extracts the EIP-2612 gas
// sponsoring info from raw payment payload JSON bytes.
func ExtractEip2612GasSponsoringInfoFromPayloadBytes(payloadBytes []byte) (*Info, error) {
	var payload struct {
		Extensions map[string]interface{} `json:"extensions"`
	}
	if err := json.Unmarshal(payloadBytes, &payload); err != nil {
		return nil, fmt.Errorf("failed to unmarshal payload: %w", err)
	}

	return ExtractEip2612GasSponsoringInfo(payload.Extensions)
}

var (
	addressPattern = regexp.MustCompile(`^0x[a-fA-F0-9]{40}$`)
	numericPattern = regexp.MustCompile(`^[0-9]+$`)
	hexPattern     = regexp.MustCompile(`^0x[a-fA-F0-9]+$`)
	versionPattern = regexp.MustCompile(`^[0-9]+(\.[0-9]+)*$`)
)

// ValidateEip2612GasSponsoringInfo validates that the EIP-2612 gas sponsoring
// info has valid format.
func ValidateEip2612GasSponsoringInfo(info *Info) bool {
	return addressPattern.MatchString(info.From) &&
		addressPattern.MatchString(info.Asset) &&
		addressPattern.MatchString(info.Spender) &&
		numericPattern.MatchString(info.Amount) &&
		numericPattern.MatchString(info.Nonce) &&
		numericPattern.MatchString(info.Deadline) &&
		hexPattern.MatchString(info.Signature) &&
		versionPattern.MatchString(info.Version)
}

File: typescript/packages/extensions/src/eip2612-gas-sponsoring/facilitator.ts
/**
 * Facilitator functions for extracting and validating EIP-2612 Gas Sponsoring extension data.
 *
 * These functions help facilitators extract the EIP-2612 permit data from payment
 * payloads and validate it before calling settleWithPermit.
 */

import type { PaymentPayload } from "@x402/core/types";
import {
  EIP2612_GAS_SPONSORING,
  type Eip2612GasSponsoringInfo,
  type Eip2612GasSponsoringExtension,
} from "./types";

/**
 * Extracts the EIP-2612 gas sponsoring info from a payment payload's extensions.
 *
 * Returns the info if the extension is present and contains the required client-populated
 * fields (from, asset, spender, amount, nonce, deadline, signature, version).
 *
 * @param paymentPayload - The payment payload to extract from
 * @returns The EIP-2612 gas sponsoring info, or null if not present
 */
export function extractEip2612GasSponsoringInfo(
  paymentPayload: PaymentPayload,
): Eip2612GasSponsoringInfo | null {
  if (!paymentPayload.extensions) {
    return null;
  }

  const extension = paymentPayload.extensions[EIP2612_GAS_SPONSORING.key] as
    | Eip2612GasSponsoringExtension
    | undefined;

  if (!extension?.info) {
    return null;
  }

  const info = extension.info as Record<string, unknown>;

  // Check that the client has populated the required fields
  if (
    !info.from ||
    !info.asset ||
    !info.spender ||
    !info.amount ||
    !info.nonce ||
    !info.deadline ||
    !info.signature ||
    !info.version
  ) {
    return null;
  }

  return info as unknown as Eip2612GasSponsoringInfo;
}

/**
 * Validates that the EIP-2612 gas sponsoring info has valid format.
 *
 * Performs basic validation on the info fields:
 * - Addresses are valid hex (0x + 40 hex chars)
 * - Amount, nonce, deadline are numeric strings
 * - Signature is a hex string
 * - Version is a numeric version string
 *
 * @param info - The EIP-2612 gas sponsoring info to validate
 * @returns True if the info is valid, false otherwise
 */
export function validateEip2612GasSponsoringInfo(info: Eip2612GasSponsoringInfo): boolean {
  const addressPattern = /^0x[a-fA-F0-9]{40}$/;
  const numericPattern = /^[0-9]+$/;
  const hexPattern = /^0x[a-fA-F0-9]+$/;
  const versionPattern = /^[0-9]+(\.[0-9]+)*$/;

  return (
    addressPattern.test(info.from) &&
    addressPattern.test(info.asset) &&
    addressPattern.test(info.spender) &&
    numericPattern.test(info.amount) &&
    numericPattern.test(info.nonce) &&
    numericPattern.test(info.deadline) &&
    hexPattern.test(info.signature) &&
    versionPattern.test(info.version)
  );
}
